lms/djangoapps/experiments/flags.py

```diff
@@ -89,19 +127,11 @@ def get_bucket(self, course_key=None, track=True):
 
         # Check if the enrollment should even be considered (if it started before the experiment wants, we ignore)
         if course_key and self.experiment_id is not None:
-            start_val = ExperimentKeyValue.objects.filter(experiment_id=self.experiment_id, key='enrollment_start')
-            if start_val:
-                try:
-                    start_date = dateutil.parser.parse(start_val.first().value).replace(tzinfo=pytz.UTC)
-                except ValueError:
-                    log.exception('Could not parse enrollment start date for experiment %d', self.experiment_id)
-                    return self._cache_bucket(experiment_name, 0)
-                enrollment = CourseEnrollment.get_enrollment(request.user, course_key)
-                # Only bail if they have an enrollment and it's old -- if they don't have an enrollment, we want to do
-                # normal bucketing -- consider the case where the experiment has bits that show before you enroll. We
-                # want to keep your bucketing stable before and after you do enroll.
-                if enrollment and enrollment.created < start_date:
-                    return self._cache_bucket(experiment_name, 0)
+            values = ExperimentKeyValue.objects.filter(experiment_id=self.experiment_id).values('key', 'value')
+            values = {pair['key']: pair['value'] for pair in values}
+
+            if not self._is_enrollment_inside_date_bounds(values, request.user, course_key):
+                return self._cache_bucket(experiment_name, 0)
```

Comment:
So, I actually think that before the start of the experiment, everyone should go to control, and after the end of the experiment, everyone should go to treatment.

That way, assuming the experiment works, we move forward correctly. If the experiment fails, then we should be removing all the code anyway.

Reply:
If the experiment fails, we may not know that for a few weeks. Like, the enrollment date bounds is so we get enough people in the experiment, and then we watch them for a while doing the course. It's not like we know at the end of the enrollment_end whether the experiment succeeded or not.

We may want to assume that it will succeed... But I'm not sure I like that. Feels too optimistic to be the ExperimentWaffleFlag default behavior.

Reply:
So, I think the other reason I wanted it to fall into giving everyone dates is so that we didn't have to maintain this holdback/bucketing code forever. If we make them drop into treatment at the end date, then we can set the end date sufficiently far out to give ourselves analysis time, and then just keep the holdback code open until enough of the hold-back users have left, and then finally delete this code. But if we drop them into control, then we'll have another group of users that we'll have to track with another chunk of code.

Would it be over-engineered to make the choice of treatment or control after the experiment ends a parameter on the flag? Or maybe I'm just overthinking the concern of having users that get the control experience for some amount of time while we wait for the data to mature, and then having them suddenly get the new experience when we get rid of the experiment gate.

Reply:
That concern doesn't bug me -- that's the experience every learner has for every other change we make without a flag.  :)   And it still assumes that the experiment is a success. The whiplash UX you worry about would still happen if we decide the experiment is a failure.

So are you suggesting that after the experiment date, everyone on the site gets the new treatment, even if the experiment was originally limited to a subset of courses or users?

I suppose also there's a risk that the engineering team has less incentive/priority in cleaning up the flag afterward.

Reply:
I think given the nature of experiments, it would make sense for people to be in the control group after we finish our bucketing for the experiment. That is our default experience until the experiment succeeds and is fully implemented and released to all users. I don't like the idea of changing the default experience while we are analyzing. Feels against the spirit of experimentation.